cayleypy/algo/random_walks.py

```diff
@@ -0,0 +1,140 @@
+            We also remove states from current layer if they appeared on some previous layer (so this also can be
+            called "non-backtracking random walk").
+            All states in the output are unique. ``y`` still can be overestimated, but it will be closer to the true
+            distance than in "classic" mode. Size of output is ``<= width*length``.
+            If ``width`` and ``length`` are large enough (``width`` at least as large as largest BFS layer, and
+            ``length >= diameter``), this will return all states and true distances to the start state.
+
+        :param width: Number of random walks to generate.
+        :param length: Length of each random walk.
+        :param start_state: State from which to start random walk. Defaults to the central state.
+        :param mode: Type of random walk (see above). Defaults to "classic".
+        :return: Pair of tensors ``x, y``. ``x`` contains states. ``y[i]`` is the estimated distance from start state
+          to state ``x[i]``.
+        """
+        start_state = self.graph.encode_states(start_state or self.graph.central_state)
+        if mode == "classic":
+            return self._random_walks_classic(width, length, start_state)
+        elif mode == "bfs":
+            return self._random_walks_bfs(width, length, start_state)
+        else:
+            raise ValueError("Unknown mode:", mode)
+
+    def _random_walks_classic(
+        self, width: int, length: int, start_state: torch.Tensor
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        """Generate classic random walks.
+
+        :param width: Number of random walks to generate
+        :param length: Length of each random walk
+        :param start_state: Starting state for all walks
+        :return: Tuple of (states, distances)
+        """
+        # Allocate memory.
+        x_shape = (width * length, self.graph.encoded_state_size)
+        x = torch.zeros(x_shape, device=self.device, dtype=torch.int64)
+        y = torch.zeros(width * length, device=self.device, dtype=torch.int32)
+
+        # First state in each walk is the start state.
+        x[:width, :] = start_state.reshape((-1,))
+        y[:width] = 0
+
+        # Main loop.
+        for i_step in range(1, length):
+            y[i_step * width : (i_step + 1) * width] = i_step
+            gen_idx = torch.randint(0, self.definition.n_generators, (width,), device=self.device)
+            src = x[(i_step - 1) * width : i_step * width, :]
+            dst = x[i_step * width : (i_step + 1) * width, :]
+            for j in range(self.definition.n_generators):
+                # Go to next state for walks where we chose to use j-th generator on this step.
+                mask = gen_idx == j
+                prev_states = src[mask, :]
+                next_states = torch.zeros_like(prev_states)
+                self.graph._apply_generator_batched(j, prev_states, next_states)
```

Comment:
Make _apply_generator_batched and _get_unique_states public (by removing the leading underscore).